### ISSUE TITLE
Loss is not converging

### ISSUE THREAD

**Issue by mbronckers** [mbronckers/GI-PVI#1]
**Issue: loss is not converging**

**Problems**

- KL divergence is increasing across epochs instead of decreasing
- The change in the pseudo-likelihood precision of the client, here "Δts.layer0_client0_nz", is not changing at all (0 matrix)

<img width="1887" alt="image" src="https://user-images.githubusercontent.com/25270093/170892808-ab0a8e88-2b9f-4ec0-a64c-4a72f53d3beb.png">

**Current findings**
-  Δts.layer2_client0_yz, Δzs.client0_z are changing across epochs
- The inducing points and associated pseudo-observations (of the last layer) are actually diverging from the training data, which is consistent with the observation that the KL divergence is increasing.

<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/170893082-962b3e6f-d994-4881-a38c-194a9a7e7b0c.png">

<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/170893096-3acd7544-e4f3-4337-bfee-37d2e385ad0e.png">

<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/170892821-06fda730-785e-48c9-9506-59bf21c8a586.png">

**Comment by mbronckers** [comment]
**Observations/questions to investigate**:


- `key, w  = q.sample(key)` results in weights of shape `[S, Dout, Din]`  but old comment said `[S, Din, Dout]`.
-  KL divergence (`kl_qp = q.kl(p)`) returns shape `[S, Dout]` but all values are equal. => this results in a KL divergence that is equal across inference samples S
<img width="948" alt="image" src="https://user-images.githubusercontent.com/25270093/170978179-44d81913-8a4d-4b01-bc63-4898242d8d6f.png">

This makes me wonder whether the inference samples S are actually different from each other.

**Comment by mbronckers** [comment]
> This makes me wonder whether the inference samples S are actually different from each other.
- The weight samples `w` of shape `[S, Dout, Din]` are different across inference samples S. (Tested via `torch.all(torch.eq(sample[0], sample[1]))`)

- But why is the log determinant of shape [2, Din] = `ratio^T @ ratio `, where ratio has shape `[S, *D_latent, Din]` equal to the same value across its shape?

<img width="880" alt="image" src="https://user-images.githubusercontent.com/25270093/170984508-e35aa105-6333-4802-a8c8-d6034aa589e4.png">

**Comment by mbronckers** [comment]
> KL divergence (kl_qp = q.kl(p)) returns shape [S, Dout] but all values are equal. => this results in a KL divergence that is equal across inference samples S

After some digging, this is only true for `layer0`. Don't know yet why this is.

**Comment by mbronckers** [comment]
After bug fixing wrt the tiling of `_zs` and the posterior construction of `q` via the likelihood factors, the client-local inducing inputs and pseudo outputs finally seem to converge but the locations are off.

<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/171024445-baf5761b-7a75-48d6-810d-a8c067c43337.png">

<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/171024416-5dfe9d73-c883-4226-bfc1-a4c7e90a3462.png">

**Comment by mbronckers** [comment]
For some reason, current initialization of inducing points (to the first training batch) is only to the first half of the training data `x_tr`:
<img width="300" alt="image" src="https://user-images.githubusercontent.com/25270093/171154648-1eb41045-5357-423d-839b-86b978767f0d.png">
Fixed this ^.

**Comment by mbronckers** [comment]
**Fixes**:
- Fixed the initialization of inducing point
- Fixed the initialization of the likelihood factor
- Fixed plotting of the variational parameters & the training data 

**Current problem**: the variational distribution moves away from the observed points.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/25270093/171176019-22330cdc-4cce-4b3d-bdbf-cd6d74b418f9.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/25270093/171176054-60786e60-07ac-4de8-8e48-5087bbea40ea.png">

**Comment by mbronckers** [comment]
> Investigate why second batch of `out` from `model.propagate(x)` is all 0

**Fix**: non-linearity() was being applied also in the final layer.

**Result**: this leads to different (better?) variational parameter optimization behavior.

<img width="350" alt="image" src="https://user-images.githubusercontent.com/25270093/171230514-5b50e000-8aa3-4485-9eb8-4c97233f0737.png">

<img width="350" alt="image" src="https://user-images.githubusercontent.com/25270093/171230514-5b50e000-8aa3-4485-9eb8-4c97233f0737.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/25270093/171232105-a600a6f9-6a34-4508-84d7-aafe9bc7f677.png">

<img width="1300" alt="image" src="https://user-images.githubusercontent.com/25270093/171232988-d916bd32-5ae2-4d1e-aa8b-e9c2932bd0e6.png">


@MattAshman ^ most recent

**Comment by mbronckers** [comment]
Matt's input:
> "I’ve also noticed that when we call `vs.requires_grad(True, *vs.names)` the variables contained in the ts no longer require gradients (I believe this is because for bounded variables Vars applies a transformation and creates a new variable, which gradients are then required for, not the original variable we passed it)
> A workaround would be to add a construction function at the beginning of estimate ELBO which build the clients using the current states of vs

See issue #2

**Comment by mbronckers** [comment]
As you can see, the variable in ts does not have require_grad = True

<img width="618" alt="image" src="https://user-images.githubusercontent.com/25270093/171275332-499b34a5-5796-4253-900a-b492a4ea3ae4.png">

**Comment by mbronckers** [comment]
Having fixed #2, it now appears that this issue has also been resolved. The ELBO curve converges and the variational parameters move slightly away from training data due to the KL term (prior is a N(0,1)).

<img width="500" alt="image" src="https://user-images.githubusercontent.com/25270093/171579206-bb0e421d-97de-46d4-ad08-808c9136fae8.png">

<img width="1235" alt="image" src="https://user-images.githubusercontent.com/25270093/171579284-0d16bac3-623e-4ecd-8df2-a1e057a2c65d.png">
